File: armorpaint/assets/plugins/viewport_celshade.js

let plugin = plugin_create();

// Register custom viewport shader
context_set_viewport_shader(function(shader) {
	// node_shader_add_uniform(shader, "vec3 light_dir", "_light_dir");
	node_shader_write(shader, " \
		vec3 light_dir = vec3(0.5, 0.5, -0.5);\
		float dotnl = max(dot(n, light_dir), 0.0); \
		output_color = basecol * step(0.5, dotnl) + basecol; \
	");
});

plugin_notify_on_delete(plugin, function() {
	context_set_viewport_shader(null);
});

File: paint/assets/plugins/viewport_celshade.c
#include "global.h"

void *plugin;

// Register custom viewport shader
void viewport_shader(void *shader) {
	node_shader_write_frag(shader, " \
		var light_dir: float3 = float3(0.5, 0.5, -0.5);\
		var dotnl: float = max(dot(n, light_dir), 0.0); \
		output_color = basecol * step(0.5, dotnl) + basecol; \
	");
};

void on_delete() {
	context_set_viewport_shader(NULL);
}

void main() {
	plugin = plugin_create();
	gc_root(plugin);

	context_set_viewport_shader(viewport_shader);
	plugin_notify_on_delete(plugin, on_delete);
}
